persist: skeleton of new multi-writer persist interface

danhhz opened the pull request:
This is an attempt to redesign the persist API to balance all of the
following factors:
- Usage within STORAGE
- Operational simplicity
- Maintainability of implementation
- Performance
- Cost

Additionally, we're also taking this shift in product priorities (as
well as some assumptions that have changed since persist started) to lay
out a plan to incrementally ship this thing to production. Shipping
persist incrementally is something everyone wants, is much less risky,
and is an earlier timeline. This is something we've tried hard to do
historically, but have been unable to tease apart into a workable plan
(given the constraints of our initial goal of fast restarts for kafka
source persistence). Without over-promising, this interface rethink
seems at the moment like it may unlock incremental delivery.

That's the good news. The bonus good news is that persist's internal
implementation is pretty clean and contains lots of great building
blocks. Even if we entirely throw out the current API and switch to this
one, it should be a pretty fast process. This is in no way a rewrite.

Here's my rough short-to-medium-term plan:
- Figure out the balance of persist priorities (support platform vs the
  kafka source fast restart alpha) with product. Adjust the following
  steps accordingly. Timeline: ASAP
- Get persist and storage to agree on this API boundary between them,
  including performance and cost expectations. Then, get sign-off from
  Cloud on the operational implications. Merge it. Timeline: ASAP
- Write and merge a prototype-quality implementation of this API to
  prove the concept and unblock storage prototyping. Timeline: ASAP
- Write and merge a design doc for the production-quality implementation
  of this API and its incremental rollout. Timeline: ASAP
- Write a bare-bones production-quality implementation of this API. This
  will happen in parallel with the initial production-quality
  implementation of STORAGE. Adjust the API as necessary in light of the
  inevitable discovery of unforeseen issues+requirements as persist and
  STORAGE are implemented. Timeline: Platform launch.
- Continue to incrementally ship persist features as decided by product
  prioritization.

### Motivation

  * This PR adds a known-desirable feature.

danhhz commented on src/persist-client/src/lib.rs:
```diff
@@ -0,0 +1,315 @@
+    _phantom: PhantomData<()>,
+}
+
+impl Client {
```
Notable departures from the current persist API (for explicit discussion):
- Write and Seal are replaced by WriteBatch. This potentially makes things much harder to use multi-threaded, but makes some other things much easier. It seems to be the way the wind is blowing.
- AllowCompaction is replaced by read capabilities.
- Persist collections no longer have String external names, only an opaque identifier assigned at creation time.
- Multiple concurrent writers are supported for a single collection in the way outlined in Petros's doc.
- Only an "async" API is offered.
- Persist no longer requires write exclusivity. The normal deployment of the "controller" is that one is running, but it's safe to start a new one while the old one is in an unknown state (distributed systems amirite). In exchange, persist depends on a TBD distributed consensus service (etcd, cockroachdb, some aws thing).

danhhz commented on src/persist-client/src/lib.rs:
```diff
@@ -0,0 +1,315 @@
+    // TODO: Bound the write and space amplification as a function of how much
+    // data has even been written to this collection as well as how much of it
+    // is "live" (doesn't consolidate out at the collection frontier).
+    pub async fn write_batch<I: IntoIterator<Item = ((K, V), T, D)>>(
```
WIP: Document latency expectations. Don't focus on this too much but an initial guess is something like "p50 < 100ms but p95/p99 > 1s". However, throughput here should be "nearly saturate your network" good. Now that I bring this up, I wonder if it's worth bringing over the current-persist separation of "received your write, ready for your next write" vs "your write is durable" (concretely: getting a PFuture from StreamWriteHandle::write is the first, the PFuture resolving a value is the second). We'll certainly need something like that to maximize throughput.

danhhz commented on src/persist-client/src/lib.rs:
```diff
@@ -0,0 +1,315 @@
+// WIP This probably doesn't need to be Clone, so I've omitted it for now.
+// Pretty sure we could make that work if necessary.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct SnapshotShard(PhantomData<()>);
```
ShapshotToken? SnapshotShardToken?

danhhz commented on src/persist-client/src/lib.rs:
```diff
@@ -0,0 +1,315 @@
+    pub async fn snapshot(
+        &self,
+        as_of: Antichain<T>,
+        num_shards: NonZeroUsize,
```
num_shards is likely either num processes or num workers. your choice!

danhhz commented on src/persist-client/src/lib.rs:
```diff
@@ -0,0 +1,315 @@
+    // be roundtrip-able as 8 bytes, persist can do some nice performance things
+    // in it's internal columnar storage format.
+    //
+    // WIP: An alternative is to separate out this encode and decode into a
```
After sitting with this, I'm starting to lean heavily toward the Codec64 thing mentioned in this WIP

danhhz commented on src/persist-client/src/lib.rs:
```diff
@@ -0,0 +1,315 @@
+pub trait DurableTimestamp: Timestamp {
+    // WIP: Figure out the requirements here. E.g. if we can require timestamps
+    // be roundtrip-able as 8 bytes, persist can do some nice performance things
+    // in it's internal columnar storage format.
```
"its" 🙈 , ditto for the copy-pasta of this comment below

danhhz commented on src/persist-client/src/lib.rs:
```diff
@@ -0,0 +1,315 @@
+#[derive(Debug, Clone, Copy, Deserialize, Serialize)]
+pub struct Id([u8; 16]);
+
+// WIP: Define the semantics of this. In particular, we should use this
```
This is something I really haven't thought enough about yet. Errors have kinda grown organically in persist so far and it's one of the messiest parts. Probably worth nailing this down at the start this time

danhhz reviewed (commented):
I have this feeling that there's a ton of thoughts I've had along the way that didn't yet make it into this document, but my brain is turning to mush for the day. This is hopefully a representative enough strawman that we can start to hammer on the edges, so please take an initial look @petrosagg @ruchirK @aljoscha

ruchirK commented:
I ran out of steam before being able to look at this today, will do so second thing tomorrow!

ruchirK commented:
Reviewing now!

ruchirK commented on src/persist-client/src/read.rs:
```diff
@@ -0,0 +1,137 @@
+        todo!("{:?}", shard);
+    }
+
+    pub async fn clone(&self) -> Result<Self, Error> {
```
is cloning semantically equivalent to someone opening another read capability on the same collection (assuming no read capabilities have called downgrade_since) or is it different? I understand cloning in the sense of "I want to give someone a copy of exactly the same read capability that I have, with visibility through the same since". 

I guess what I'm not sure about is: if someone makes a new capability, shouldn't it be from a since <= my read capability's since? (assuming single dimensional times). And so, given all of that, its not clear to me why someone would want to clone a read cap.

danhhz replied:
yeah, this is intended to be cloning as defined in the formalism doc, where my understanding is that a cloned read handle gets a since that is the same as this one at cloning time but independent thereafter (ditto write handle clone and upper)

I also am in the dark on when a handle will be cloned vs obtained via something like open_collection. the formalism doc really just defines things, not how they'll be used. maybe Petros has a better sense for this?

> if someone makes a new capability, shouldn't it be from a since <= my read capability's since?

I think it would have to be >=, no? sort of beside the point, since we choose =, but I want to make sure I'm not missing something

petrosagg replied:
I'm not sure what Frank had mind but maybe it's useful for active-replication when we restart instances?

ruchirK replied:
Here's my understanding. Let's say we have two read capabilities A and B for the same collection. A has a since of 5, and B has a since of 10. My understanding is that the collection's since frontier is 5, because it has to be the min of all of the since frontiers for all of the read capabilities (in single dimensional times), because if the since frontier for the collection was say 7, that seems logically inconsistent with A's since frontier of 5. This might not be how the formalism doc thinks about this -- in which case my apologies.

In this framing, a new read capability would be instantiated at the collection's since frontier (in this case 5), which is <= the since frontier of all of the existing read capabilities (A has 5, B has 10)

danhhz replied:
oh I see. my understanding from the formalism doc is that you get the since/upper of the handle you clone, not the overall since/upper of the collection at the time you clone. I was just matching that behavior here, but we can do whatever

ruchirK replied:
yeah sorry - on clone you get the since frontier of the cap you cloned, but if you mint a new cap you get the since frontier of the collection which is <= the since frontiers of all of the existing read caps, which makes it less clear to me why you would want/need to clone

danhhz replied:
aha! now i get you. yeah I'm curious what @petrosagg  and @aljoscha  think here on when each might be used and whether we need both

ruchirK commented on src/persist-client/src/write.rs:
```diff
@@ -0,0 +1,69 @@
+        todo!()
+    }
+
+    // NB new_upper of the empty antichain "finishes" this collection, promising
```
nit: just to make sure I'm on the same page `new_upper` of the empty antichain finishes the collection for this particular write handle right? other write handles might still be writing stuff at various times?

danhhz replied:
correct, though since they're required to be writing the same thing, presumably they'll follow suit shortly. because of that contract, it does mean we know everything that will ever be written to this collection tho (we don't really do anything with at atm, just pointing the implication out)

ruchirK commented on src/persist-client/src/write.rs:
```diff
@@ -0,0 +1,69 @@
+}
+
+impl<K, V, T, D> WriteHandle<K, V, T, D> {
+    pub async fn deregister(&mut self) -> Result<(), Error> {
```
question about deregister for both reads and writes: deregister is morally the same as dropping the capability right? with the idea that you can mint a new pair of (read, write) capabilities by calling open_collection on the same collection, but when you do so the caps may have advanced arbitrarily far in the future?

danhhz replied:
yup, possibly we just want to inline the definition in impl Drop. i can't see why anyone would explicitly call this. (notably, it's not `pub` here)

petrosagg replied:
I think it's correct for this to move to Drop since this should probably be a `self` receiver instead of `&mut self` because it currently leaves the handle in an unusable state. So if it's `self` and it's going to get dropped anyway, it can just go in the Drop impl

danhhz replied:
> (notably, it's not pub here)

lol this was a lie. okay, i'll just get rid of this method

btw, any idea why drop takes `&mut self` instead of `self`? I would have originally written this as `self` but then I wouldn't have been able to call it from drop

ruchirK commented on src/persist-client/src/lib.rs:
```diff
@@ -0,0 +1,114 @@
+        todo!();
+    }
+
+    pub async fn open_collection<K, V, T, D>(
```
this api seems like it requires the user to hold onto the opaque collection id across restarts to reopen existing collections -- is that a fair understanding? I guess my followup question then is: I don't immediately see the benefit of handing back to the user an opaque token vs letting them pick a meaningful string name and then storing the string name somewhere in our distributed consensus service. The api as written seems like it would require the user to remember "ok, opaque id foo corresponds to collection <bar> for my application" and make that knowledge available to all peer writers

danhhz replied:
I think in practice it ends up being the same thing. the user either has to durably write down the opaque id or (to protect against the name construction logic changing over time) they have to durably write down the name

anyway, I don't care which we do. we can easily do either and this is 100% STORAGE's preference

petrosagg replied:
Regardless of the type of collection id (enforcing it to be a `uuid` is fine) there is a meaningful reason that persist should accept a name from the outside which is that if `persist` picks the name and then STORAGE needs to write it down there is the opportunity for orphan collections if STORAGE crashes at that point. Not a huge deal but if we can easily prevent it then why not

danhhz replied:
great point

ruchirK reviewed (commented):
Left my first round of comments! I think everything here seems reasonable I need to go back over the formalism doc and petros's storage doc to remember how the big picture is supposed to fit together

petrosagg commented on src/persist-client/src/lib.rs:
```diff
@@ -0,0 +1,114 @@
+// - Figure out the contract of tokio Runtimes and who is responsible for making
+//   sure they're in the TLC.
+// - I've split up getting a snapshot vs tailing for new changes below, because
+//   I think it simplifies the story about multi-process snapshots. We could
```
What is a multi-process snapshot?

danhhz replied:
see my response to your question on snapshot shards

petrosagg commented on src/persist-client/src/lib.rs:
```diff
@@ -0,0 +1,114 @@
+//   the formalism doc? Many of them have direct analogues down at the persist
+//   level.
+// - At the moment, we have both a STORAGE collection concept as well as a
+//   persist collection concept. One of these should be renamed :).
```
fwiw I'm happy to always prefixing the word "collection" if we have trouble finding a better name

danhhz replied:
after answering aljoscha's questions in slack today, i'm starting to wonder if we call the persist thing a "shard" and just say that a storage collection is make up of 1 or more persist shards. wdyt?

petrosagg commented on src/persist-client/src/lib.rs:
```diff
@@ -0,0 +1,114 @@
+//   we should think about how this will be used.
+// - Leases for read and write handles, as described in formalism.
+// - I think some things (readers at least?) will want to connect to the persist
+//   controller over rpc. At the very least, this is something that we may want
```
Is the persist controller a singleton? My understanding was that we wanted to have many of them and have them publish a small amount of into to etcd (or equivalent)

petrosagg commented on src/persist-client/src/lib.rs:
```diff
@@ -0,0 +1,114 @@
+
+    pub async fn new_collection<K, V, T, D>(
+        &self,
+    ) -> Result<(ReadHandle<K, V, T, D>, WriteHandle<K, V, T, D>), Error>
```
This should either accept or return a collection id, right?

danhhz replied:
at the moment i have it assigning a random one, but this is exactly the sort of thing i'm looking to hash out in this PR. I think your comment in the other thread about the collection id being orphaned is a good one.

Q back at you: if this takes an id, then the impl is basically exactly the same as open_collection, except that this one can potentially guarantee that the id isn't already in use and that the since and uppers of the returned handles are both `Antichain::from_elem(T::minimum())`. do you think the distinction is worth keeping? persist can do whatever here, it's really just what's best for storage

petrosagg commented on src/persist-client/src/read.rs:
```diff
@@ -0,0 +1,137 @@
+    // WIP: The impl is far far easier if this is something we keep single
+    // process at first (as opposed to being able to shard it like snapshots).
+    // Do we think that's fine for short term (i.e. initial platform launch)? We
+    // can definitely revisit this later if it ends up being a bottleneck.
```
I'm not sure I follow this. What are the snapshot shards and what is being kept single process? The way I understanding reading from a persist collection in a mz platform setting is that each persist collection will be read by at most one worker per process.

danhhz replied:
also responding to the following comment

> What is a snapshot shard? A STORAGE collection would be composed of renditions which have a specific number of shard. Each rendition shard is backed by a single persist collection.
>
> When a STORAGE collection is replayed it will be replayed to as many workers as there are shards. If it makes your life easier we don't need persist itself to figure out how to shard a persist collection to an arbitrary number of readers

oof I've confused things by reusing the word shard here, it's totally independent of rendition shards. pretend I used "snapshot part" everywhere instead. a single reader per persist collection is fine by me and certainly simpler

the two cases I was thinking about that led me to think about this are:
- adding a tiny bit of independence to the SOURCE+COMPUTE boundary or the COMPUTE+COMPUTE boundary. for example, a source could write a STORAGE COLLECTION with whatever number of shards N is most natural (one per kafka partition, perhaps). then in COMPUTE, we could distribute work of the N snapshot replays each across M workers. if N=M it's straightforward to do what you're saying, but requiring it at every level of the system seems like it will mean far more renditions as things have to get resharded (which, like, they're not free). ditto this same idea if we want to chain two CLUSTERs together using a STORAGE collection between them
- extension to the first thing, making this also work with the timely dataflow feature where workers in a single dataflow can be spead across separate processes. last time I asked, nikhil said this won't be part of platform v1 but we might use it later. dunno

thoughts on those concerns?

petrosagg replied:
Just to make sure we're on the same page, my thought was that STORAGE provides a collection with some number of shards, and then any COMPUTE cluster, with however many workers, can read that collection by assigning the duty of reading a particular storage shard to a worker. If the cluster has more workers than storage shards some workers will be idle for that first part of ingestion but an compute level exchange can fix that.

if it's easy for persist provide a sharded reader then that sounds like a fine optimization to have.

petrosagg commented on src/persist-client/src/read.rs:
```diff
@@ -0,0 +1,137 @@
+    pub async fn snapshot(
+        &self,
+        as_of: Antichain<T>,
+        num_shards: NonZeroUsize,
```
What is a snapshot shard? A STORAGE collection would be composed of renditions which have a specific number of shard. Each rendition shard is backed by a single persist collection.

When a STORAGE collection is replayed it will be replayed to as many workers as there are shards. If it makes your life easier we don't need `persist` itself to figure out how to shard a persist collection to an arbitrary number of readers

danhhz replied:
see my response to your other question on snapshot shards

petrosagg commented on src/persist-client/src/write.rs:
```diff
@@ -0,0 +1,69 @@
+    // TODO: Bound the write and space amplification as a function of how much
+    // data has even been written to this collection as well as how much of it
+    // is "live" (doesn't consolidate out at the collection frontier).
+    pub async fn write_batch<I: IntoIterator<Item = ((K, V), T, D)>>(
```
Not to get into the weeds here, but it may make sense to have these be references since I assume they will be encoded into a byte buffer and not used directly. Correct me if I'm wrong

danhhz replied:
no you're absolutely right. we have a couple things like this now and so far have just ended up using the owned version so the reference one was a usability papercut, but I agree that if we're only going to have one, it should be the one that makes better perf possible. is this what you're thinking?

> <I: IntoIterator<Item = ((&'_ K, &'_ V), &'_ T, &'_ D)>>

if not, just type what you want and we can make it work

petrosagg commented on src/persist-types/src/lib.rs:
```diff
@@ -50,3 +50,34 @@ pub trait Codec: Sized + 'static {
+//
+// WIP: Figure out the requirements here. E.g. if we can require timestamps and
+// diffs be roundtrip-able as 8 bytes, persist can do some nice performance
+// things in its internal columnar storage format compared to if they were tied
```
For my curiosity, what can we optimize if ts and diffs serialize to 64 bits? I can't remember exactly (I can find the slack thread if useful) but we had one usecase where we wanted to persist a diff that was something complicated and we ended up doing a bunch of work to make look like a collection because the diffs were hardcoded to `isize`.

Could whatever optimization we get from this being 8bytes implemented by having a `hint` method on the generic `Codec` trait?

danhhz replied:
two things:
- it's a columnar format, so if you have fixed sized things (and the schema knows they're fixed size) then you can avoid encoding the lengths of them
- inside parquet (the concrete columnar format we use), a bunch of of the compression algorithms are only available for columns of integral types and not for columns of byte arrays

> Could whatever optimization we get from this being 8bytes implemented by having a hint method on the generic Codec trait?

I don't think but perhaps I'm missing something

aljoscha commented:
Mostly requests for clarification so far. :sweat_smile: 

What is the thing that allows us to have multiple writers with this API? I'm guessing it's the combination of `write`/`seal` into one method, but I don't yet understand why. What type of _multiple writer_ situation do we want to support? Is it "multiple writers write exactly the same data", as mentioned in the STORAGE document? Or is it "multiple writers write the data of their partition? I can imagine the latter being useful for the case where you have one persisted collection backing a Kafka source, and you have one process writing updates from partition 0 and another process writing updates from partition 1, but to the same collection.

Can `Client::new()` be called from multiple processes for a given location? I'm assuming yes but wanted to double check.

Do the shards we see in the snapshot API correspond to shards as described in the STORAGE doc or are they rather the mechanism that we use to get parallel snapshot reading. I'm assuming the latter, but am not very sure. If yes, then the naming might be confusin between the two docs. (cc @petrosagg).

I have yet to think about how a source implementation would use the API. One thing that obviously wouldn't work anymore is independent writing and sealing, which we use in the current persistent source implementation to seal the different collections that are backing a source atomically.

danhhz reviewed (commented):
thanks for the comments! I think I've responded to all of them, will push updates to the PR async

> What is the thing that allows us to have multiple writers with this API? I'm guessing it's the combination of write/seal into one method, but I don't yet understand why. What type of multiple writer situation do we want to support? Is it "multiple writers write exactly the same data", as mentioned in the STORAGE document? Or is it "multiple writers write the data of their partition? I can imagine the latter being useful for the case where you have one persisted collection backing a Kafka source, and you have one process writing updates from partition 0 and another process writing updates from partition 1, but to the same collection.

yeah, it's exactly "multiple writers write exactly the same data", as mentioned in the STORAGE document and #10913

> Can Client::new() be called from multiple processes for a given location? I'm assuming yes but wanted to double check.

yep!

> Do the shards we see in the snapshot API correspond to shards as described in the STORAGE doc or are they rather the mechanism that we use to get parallel snapshot reading. I'm assuming the latter, but am not very sure. If yes, then the naming might be confusin between the two docs. (cc @petrosagg).

different things. the naming clash is my fault, I'll update. also see my response to petros on snapshot shards

> I have yet to think about how a source implementation would use the API. One thing that obviously wouldn't work anymore is independent writing and sealing, which we use in the current persistent source implementation to seal the different collections that are backing a source atomically.

yeah, this is definitely something i've love for you and petros to think through as soon as possible :)

danhhz commented:
Added a first swing at rustdoc. I think I caught all the changes I promised, please let me know if I missed any. Also renamed "tail" to "listen" given that Frank was just renaming some things that were both called "tail" at the coord/dataflow boundary

aljoscha commented:
I think one of my questions got lost: What is the thing that allows us to have multiple writers with this API, or why wouldn't the old API allow that? I'm guessing it's the combination of write/seal into one method, but I don't yet understand why.

danhhz commented:
> I think one of my questions got lost: What is the thing that allows us to have multiple writers with this API, or why wouldn't the old API allow that? I'm guessing it's the combination of write/seal into one method, but I don't yet understand why.

I tried to answer it but perhaps didn't do a good enough job :). Here's more words in the hope that helps. There's two things here.

One is the concept introduced in https://github.com/MaterializeInc/materialize/pull/10913. Multiple writers of the same collection are required to be writing the same data (same in the sense of definiteness), so we're free to use whichever we get first. Each writer places batches in s3 in a unique sub-dir called "pending". The main persist state machine finds out about these, either by watching or rpc, and plugs them in as inputs to the machine in Description order. The combination of write and seal into one method isn't necessary for this, just the semantics of definiteness. (A followup optimization that is allowed but not required is that readers can duplicate just enough of the state machine logic to peek ahead at data in "pending" and use it.

I think that's the one you're asking about but just in case, here's the other. Previously, we required each "storage location" to have one exclusive writer, which was the persist state machine. Now, we tolerate multiple ones (to make an easier operational story) by having them each operate in unique sub-dirs and using a tiny bit of distributed consensus (via an external service) to handle flipping from one version of META to the next. This will allow cloud to simply restart persist coordinators that are in an unknown state, and if they come back, one of the two discovers this and bails.

(All of this btw, is very much "proposal for discussion and validation" not "decision that was made while you were on vacation".)

danhhz commented on src/persist-client/src/write.rs:
```diff
@@ -0,0 +1,67 @@
+    /// (in the sense of "definite"-ness).
+    pub async fn write_batch<'a, I: IntoIterator<Item = ((&'a K, &'a V), &'a T, &'a D)>>(
+        &mut self,
+        updates: I,
```
discussed in a 1:1 with petros today that this may want to be an async iterator for things like a big s3 source object being introduced all at one time

danhhz commented on src/persist-client/src/write.rs:
```diff
@@ -0,0 +1,67 @@
+        todo!()
+    }
+
+    /// Applies `updates` to this collection and downgrades this handle's upper
```
reminder to self: document that this uses a bounded amount of memory, even when the input iterator is very large. individual records are the more granular this gets though, so they each need to be a sane size

guswynn commented on src/persist-client/src/read.rs:
```diff
@@ -0,0 +1,149 @@
+    // bandwidth. My thinking so far is that Map and Filter should stay
+    // STORAGE-level concerns.
+    //
+    // WIP: My first attempt here was to return `impl Stream` but that didn't
+    // compile with `todo!`. I don't yet have an opinion on which is better.
```
if you do want to do that, https://docs.rs/futures-util/latest/futures_util/stream/fn.empty.html would work!

danhhz commented on src/persist-client/src/lib.rs:
```diff
@@ -0,0 +1,145 @@
+//   about.
+
+// WIP
+// - Which of these need to be durably storable (for restarts) or transmittable
```
seems like `Id`, `Location`, and `ReadHandle` need to be able to go over the network

danhhz commented on src/persist-client/src/lib.rs:
```diff
@@ -0,0 +1,145 @@
+// - Same question for clone-able. The formalism document describes the
+//   storage-level read and write capabilities as being clone-able, but does
+//   this translate to a persist requirement as well?
+// - Figure out the contract of tokio Runtimes and who is responsible for making
```
Caller puts it in TLC and we document which methods need that

danhhz commented on src/persist-client/src/lib.rs:
```diff
@@ -0,0 +1,145 @@
+//   this translate to a persist requirement as well?
+// - Figure out the contract of tokio Runtimes and who is responsible for making
+//   sure they're in the TLC.
+// - I've split up getting a snapshot vs listening for new changes below,
```
Seems fine

danhhz commented on src/persist-client/src/lib.rs:
```diff
@@ -0,0 +1,145 @@
+//   because I think it simplifies the story about multi-process snapshots. We
+//   could also instead do something more analogous to SubscribeAt from the
+//   formalism doc.
+// - We'll want to expose overall collection frontier information (since and
```
Punt for now

danhhz commented on src/persist-client/src/lib.rs:
```diff
@@ -0,0 +1,145 @@
+//   upper) at the very least for introspection. There's a few options here so
+//   we should think about how this will be used.
+// - Leases for read and write handles, as described in formalism.
+// - To what extent should this adopt the naming conventions for STORAGE used in
```
Let's keep what we have now

danhhz commented on src/persist-client/src/read.rs:
```diff
@@ -0,0 +1,149 @@
+    /// downgrade their read capability when they are certain they have all data
+    /// through the frontier they would downgrade to.
+    //
+    // WIP: This is an opportunity to push down Projection information to save
```
two classes of mfps: one is create source level, one is usage level. initially seems fine that persist will only deal with project, but neither is doing anything with MFP for v0. on second thought, filter might be useful for keys

danhhz commented on src/persist-client/src/read.rs:
```diff
@@ -0,0 +1,149 @@
+    // WIP: Should this live on ReadHandle or somewhere else? E.g. I think it
+    // could also live on Client.
+    //
+    // WIP: Figure out the error story here. Retries should move inside this, but
```
Retry-er must be observable. Petros's initial reaction is to expose an error on whether it's retryable

danhhz commented:
Okay @aljoscha @petrosagg @ruchirK, this is ready for a look! As discussed, hoping to merge this early next week to help enable storage prototyping

(Let me know if I missed anything, please! My brain is pretty fried so I wasn't as careful as I normally am about making sure that everything got addressed)

aljoscha commented on src/persist-client/Cargo.toml:
```diff
@@ -0,0 +1,19 @@
+description = "Client for Materialize pTVC durability system"
+version = "0.0.0"
+edition = "2021"
+rust-version = "1.58.0"
```
Should be "1.59.0" by now.

danhhz replied:
done

aljoscha commented on src/persist-client/src/read.rs:
```diff
@@ -0,0 +1,224 @@
+    }
+}
+
+/// A token representing one partition of a "snapshot" (the contents of a shard
```
nit: sorry for being somewhat late with this, but: I don't know if "partition" is a good term for this. When I hear partition I think of things like hash-partitioning, which is related to sharding (and that term is also used already). Maybe _split_ is a good term for this thing? To me, it has fewer connotations of a "purposeful" partitioning of the data.

Feel free to ignore, though.

danhhz replied:
split works for me!

aljoscha commented on src/persist-client/src/read.rs:
```diff
@@ -0,0 +1,224 @@
+    /// read at times not greater or equal to `new_since`.
+    ///
+    /// This may trigger (asynchronous) compaction and consolidation in the
+    /// system. A `new_upper` of the empty antichain "finishes" this shard,
```
```suggestion
    /// system. A `new_since` of the empty antichain "finishes" this shard,
```

danhhz replied:
fixed

aljoscha commented on src/persist-client/src/read.rs:
```diff
@@ -0,0 +1,224 @@
+        &self,
+        timeout: Duration,
+        new_since: Antichain<T>,
+    ) -> Result<Result<(), InvalidUsage>, LocationError> {
```
I'm probably confused, but I would think that the error types should be swapped here. This would allow a caller to bubble up the `InvalidUsage` error using `?`. These usage errors always need to bubble up and stop the system because they represent bugs, I think. And then only the "real" `LocationError` remains. Could also be that I misunderstood the _sled.rs_ piece and have to go back and read it again.

danhhz replied:
good point. my reading of the sled errors post was the same as yours and that's how I started. I swapped to this because it made error handling far easier internally in the prototype (we got to use ? far more often) and the brief bit of thinking I did about it made me think the caller doesn't actually care about the order, just that they're two separate results with different error types. but I consider this unresolved, especially until we've got a bit of usage of this api under our belt, so added this as a top level TODO

aljoscha reviewed (approved):
It had some nits/questions, but please merge early because we can always resolve those later.

danhhz reviewed (commented):
I back-ported a couple final tweaks from the prototype that I missed the first time I did that. none of them feel like they should be objectionable, but in the interest of transparency, here they are:

- made ReadHandle::downgrade_since `&mut self` instead of `&self`
- added a `T: Timestamp + Lattice + Codec64` bound to ReadHandle to match WriteHandle
- swapped the order of read and write handles in the return value of Client::open
- added Serialize, Deserialize to WriterId to match ReaderId
- changed the return type of ReadHandle::listen to actually return `Listen` instead of `()` (whoops)
- made the type params of ListenEvent <K, V, T, D> to match the others

TFTRs everyone!

danhhz commented:
I have the holdover in-mem implementation mostly working but it turned out to be a tiny big larger than what I originally guessed, so I'm going to PR that as a fast follow, instead of just inlining it in this one. should be out for review later this week
